FIFO mode option for HID OutputReports

JoergAtGithub opened the pull request:
While #4585 makes the Mixxx HID subsystem fast and responsive in overload situations for any device, that use reports in the sense of the USB HID Class specification, these optimizations do not work for the Traktor Kontrol S4 Mk3, which uses HID reports to transport the data of it's own protocol.
First attemp to address this S4 specific protocol was https://github.com/mixxxdj/mixxx/pull/4692 (Option for resending reports with identical data), but this was not enough. It came out #10828, that the S4 protocol also depend on sequences of reports with the same ReportID.

This PR replaces the Resend-Option from #4692 with an Option to use a FIFO for reports, which must not be skipped.

Closes: #10828

JoergAtGithub committed:
Removed resendUnchangedReport implementation from PR #4692 form HidIoOutputReport

JoergAtGithub committed:
Removed redundant printing of device name and serial number
Nowadays the logging system prints this information at begin of each line anyway

JoergAtGithub committed:
Renamed option resendUnchangedReport to useNonSkippingFIFO and described the new behavior

JoergAtGithub committed:
Added implementation of FIFO for OutputReports

JoergAtGithub committed:
Use the new OutputReport FIFO as alternative mode in HidIoThread

uklotzde commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    // Handle the case, that the FIFO is full - which is an error case
+    if (m_indexOfLastSentReport == indexOfReportToCache) {
```
Overwriting old, unsent data with new incoming data would probably be the more approriate strategy. A lock-free FIFO wouldn't allow this, but since you are locking here anyway why not use this alternative overflow strategy?

JoergAtGithub replied:
This is a very good point and I'm unsure what's the best strategy to handle an overflow:

1. Overwriting the last report ensures, that vital initialization reports at startup will be send if the overflow happens at startup
2. Overwriting the oldest report would propably better recover, when the overflow happens during the session
   This assumens that new report data supersede previous report data, which is the case for class compliant report usage, but the FIFO mode is especially for devices that use reports in a non-class compliant way.

uklotzde commented on src/controllers/hid/hidioglobaloutputreportfifo.h:
```diff
@@ -0,0 +1,41 @@
+    /// Mutex must be locked when reading/writing
+    /// m_outputReportFifo, m_indexOfLastSentReport
+    /// or m_indexOfLastCachedReport
+    QMutex m_fifoMutex;
```
Why a mutex and not a lock-free SPSC queue with fixed capacity?

JoergAtGithub replied:
I changed the implementation to SPSCQueue now.

daschuer commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,131 @@
+    cachedData.append(data);
+
+    // Swap report data to FIFO
+    cachedData.swap(m_outputReportFifo[indexOfReportToCache]);
```
Here we allocate memory, swap the pointer to the memory and delete the old memory. 
Did you consider to reuse the memory?

JoergAtGithub replied:
Here we've a FIFO that can store reports with different ID and therefore different length. In case of identical length of old and new report, I could reuse the memory of cause.

daschuer commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,131 @@
+    QByteArray cachedData;
+    cachedData.reserve(kReportIdSize + data.size());
+    cachedData.append(reportId);
+    cachedData.append(data);
```
This is a deep copy of the QByteArray. Maybe there is a way around it? 
Not mandatory, just an idea.

JoergAtGithub replied:
We could swap the arrays and prepend the byte with he report ID. In Qt5 this is a deep-copy too, but this changes when we update to Qt6: https://doc.qt.io/qt-6/qbytearray.html#prepend

JoergAtGithub replied:
I changed the implementation to std::swap and prepend now.

acolombier commented:
Tested with #10937 and it fixes the issue!

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.4' into hid_fifo_mode3

JoergAtGithub committed:
Changed HidIoGlobalOutputReportFifo implementation to lockless SPSCQueue
Use QByteArrays prepend function, which will be a "very fast constant time" function when we switch to Qt6

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.4' into hid_fifo_mode3

uklotzde commented on src/controllers/hid/hidcontroller.h:
```diff
@@ -71,32 +71,64 @@ class HidControllerJSProxy : public ControllerJSProxy {
     /// @param length Unused but mandatory argument
     /// @param reportID 1...255 for HID devices that uses ReportIDs - or 0 for devices, which don't use ReportIDs
-    /// @param resendUnchangedReport If set, the report will also be send, if the data are unchanged since last sending
+    /// @param useNonSkippingFIFO (optional)
```
How about using an `enum class SendReportDispatchStrategy { LatestDedup, Fifo }`? Documenting the enum variants once instead of duplicating the documentation for every function would be more suitable.

JoergAtGithub replied:
AFAIK, it's not possible to access C++ enums from JavaScript code, or am I wrong here?

uklotzde replied:
I didn't consider that the signature must be compatible with JS. But simply duplicating the documentation is not maintainable.

uklotzde replied:
Or use constants (int or bool) and document those instead.

uklotzde commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,97 @@
+        const RuntimeLoggingCategory& logOutput) {
+    // First byte must always be the ReportID-Byte
+    QByteArray report(std::move(reportData));
+    report.prepend(reportId);
```
Writing the first byte and then appending the remaining data would be more efficient.

Using `std::move` on `const QByteArray&` is pointless.

JoergAtGithub replied:
> Writing the first byte and then appending the remaining data would be more efficient.

This was the old code, I changed it because the Qt6 documentation of prepend says: "This operation is typically very fast ([constant time](https://doc.qt.io/qt-6/containers.html#constant-time)), because [QByteArray](https://doc.qt.io/qt-6/qbytearray.html) preallocates extra space at the beginning of the data, so it can grow without reallocating the entire array each time."

uklotzde replied:
Ok. Maybe a short comment/link would help? If you are not a Qt expert those subtle details are surprising.

uklotzde commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,97 @@
+        return false;
+    }
+
+    auto fifoUsedEntries = m_fifoQueue.size();
```
The result of this (costly) operation is used for logging and should not be invoked otherwise. It accesses 2 atomics with acquire ordering and requires a memory fence.

Don't use lock-free data structures like any other, simple container class.

uklotzde commented on src/controllers/hid/hidiothread.cpp:
```diff
@@ -204,11 +205,30 @@ void HidIoThread::updateCachedOutputReportData(quint8 reportID,
 
+    // If useNonSkippingFIFO is false, the report data are cached here
+    // If useNonSkippingFIFO is true, this cache is cleared
     actualOutputReportIterator->second->updateCachedData(
```
Why is the mutex locked and this operation invoked even when using the FIFO strategy? The cache and the mutex should be bypassed. Ideally the dispatch strategy for each reportID would be immutable and configured at construction time. At least the FIFO report IDs need to be known in advance to avoid the locking.

JoergAtGithub replied:
The call of updateCachedData is always needed to guarantee seamless transitions between both modes.
If a skipable report is cached, it must be cleared, because otherwise it would be send after the report from the non-skipable report from the FIFO - and would overwrite with outdated data. This happens in updateCachedData as well as the check for errors.

uklotzde replied:
I acknowledge that the current design is not suitable to for this optimization. But in principle the dispatch strategy should be fixed per report ID and not change with every call.

uklotzde replied:
Caching the mode only for the sender would avoid locking the mutex with every call. The cache only needs to be cleared once after changing the mode.

JoergAtGithub replied:
The only datastructure per output report is currently HidIoOutputReport. Adding a additional data structure per output report would prevent the mutex locking in FIFO mode. But this mutex is only locked for the very short time of looking up the map for the map iterator of the report.
This takes nanoseconds, while the sending of a report takes several milliseconds. And in FIFO mode, calling this function must not be done in higher rate than the HID subsystem can send reports to the device. This ratio makes a mutex blocking very unlikely - and even when it happens, it will block the other thread only for several nanoseconds.
I don't think, that it makes sense to add additional code here, because the impact will be insignificant in practice.

uklotzde replied:
However short and tiny the critical section might be, you have no control over the OS scheduler. Generous use of mutexes is always a red herring.

JoergAtGithub committed:
Removed costly m_fifoQueue.size() operation, which was only needed for logging

JoergAtGithub committed:
Added commend about performance of QByteArrays prepend method

JoergAtGithub committed:
Removed pointless std::move on const QByteArray&

JoergAtGithub committed:
Deduplicated comments for argument useNonSkippingFIFO of send and sendOutputReport

acolombier commented:
I did some spot checks on the latest version and this still seems to work just fine.

daschuer commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,93 @@
+        return false;
+    }
+
+    QByteArray dataToSend(std::move(*pFront));
```
dataToSend has the report ID prepended. This is not clear initially. 
Can we improve the situation by always putting the report ID at the beginning? 
This will also solve the reallocation of the buffer when prepending the report ID before putting it into the buffer. 

The same benefit goes for 
void HidController::send(QByteArray data, unsigned int reportID) 
The we just not store the modified buffer.

JoergAtGithub replied:
Our Javascript API has seperated arguments for ReportID and report data. This is not only needed for backwards compatibility, but is also the cleanest way to define such an API.
I will improve naming and add a comment, to make clear that this variable contains ReportID + data.

daschuer reviewed (changes requested):
Thank you for this PR. In general it should work good, though I am not able to test. But since this is optional we can merge it anyway. 
I have added some suggestions to improve performance and readability.

Swiftb0y commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,93 @@
+        const RuntimeLoggingCategory& logOutput) {
+    // First byte must always be the ReportID-Byte
+    QByteArray report(reportData);
+    report.prepend(reportId); // In Qt6 this is a very fast operation without reallocation
```
https://github.com/qt/qtbase/blob/9e7f1f3e5707c91fcd6dd5cf6fd30b30f4b1bd58/src/corelib/text/qbytearray.cpp#L1971-L1985

JoergAtGithub replied:
The Qt6 documentation explicitly say, that this is very fast: https://doc.qt.io/qt-6/qbytearray.html#prepend

JoergAtGithub committed:
Use size_t for buffer size and string len

JoergAtGithub commented on src/controllers/hid/hidioglobaloutputreportfifo.cpp:
```diff
@@ -0,0 +1,93 @@
+} // namespace
+
+HidIoGlobalOutputReportFifo::HidIoGlobalOutputReportFifo()
+        : m_fifoQueue(kSizeOfFifoInReports) {
```
I think this is a missunderstanding. In is not a abreviation for Input here, its just the word in. The whole code is only for OutputReports.
InReports is the counted unit, like InMillis or InSeconds for times.

JoergAtGithub commented on src/controllers/hid/hidiooutputreport.cpp:
```diff
@@ -29,35 +29,41 @@ HidIoOutputReport::HidIoOutputReport(
 
     } else {
-        if (m_possiblyUnsentDataCached) {
+        if (m_possiblyUnsentDataCached && !useNonSkippingFIFO) {
```
This is correct. The member stores the data size for this ReportID.
This information must not be invalidated by clearing the cached data. Data size of a specific HID report can't change, as it's defined by the device hardware. If a mapping tries to send with a different data size to the same ReportID, this is an error which we must detect.

JoergAtGithub committed:
Use type explicit instead of auto where the type is out of sight

JoergAtGithub committed:
Clarified that array contains the ReportID

JoergAtGithub commented:
Have a look at #11284 which requires #11326 to work.

JoergAtGithub commented:
This is still ready to be merged:
- Review remarks addressed 
- That the adjusted functionality now really fix the issue (Bug #10828) was confirmed above
- Flawless operation over a longer period of time is also confirmed above

daschuer reviewed (approved):
Thank you. 
Than let's quickly add it to the 2.4-beta to have the chance for user feedback before release.

JoergAtGithub commented:
Thank you! No HID support in 2.4 is complete without known bugs. I will prepare a blogpost in the comming days.
